examples/model_selection/plot_grid_search_digits.py

```diff
@@ -13,56 +13,177 @@
-X, y = datasets.load_digits(return_X_y=True)
+digits = datasets.load_digits()
+
+# %%
+# To apply a classifier on this data, we need to flatten the images.
+# Each image of 8 by 8 pixels needs to be transformed to a vector of 64 pixels.
+# Thus, we will get a final data array of shape (n_images, n_pixels)
+n_samples = len(digits.images)
+X = digits.images.reshape((n_samples, -1))
+y = digits.target == 7
+print(
+    f"The number of images is {X.shape[0]} and each image contains {X.shape[1]} pixels"
+)
+
+# %%
+# As presented in the introduction, the data will be split into a development
+# and evaluation set of equal size.
+from sklearn.model_selection import train_test_split
 
-# Split the dataset in two equal parts
 X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=0)
 
-# Set the parameters by cross-validation
+# %%
+# Define our grid-search framework
+# --------------------------------
+#
+# We will fine-tune a classifier by searching the best hyper-parameters on the
+# development dataset. We need to define the values of the hyper-parameters
+# and the scores to compute to later on select the best candidate.
+
 tuned_parameters = [
     {"kernel": ["rbf"], "gamma": [1e-3, 1e-4], "C": [1, 10, 100, 1000]},
     {"kernel": ["linear"], "C": [1, 10, 100, 1000]},
 ]
 
 scores = ["precision", "recall"]
 
-for score in scores:
-    print("# Tuning hyper-parameters for %s" % score)
-    print()
-
-    clf = GridSearchCV(SVC(), tuned_parameters, scoring="%s_macro" % score)
-    clf.fit(X_train, y_train)
-
-    print("Best parameters set found on development set:")
-    print()
-    print(clf.best_params_)
-    print()
-    print("Grid scores on development set:")
-    print()
-    means = clf.cv_results_["mean_test_score"]
-    stds = clf.cv_results_["std_test_score"]
-    for mean, std, params in zip(means, stds, clf.cv_results_["params"]):
-        print("%0.3f (+/-%0.03f) for %r" % (mean, std * 2, params))
-    print()
-
-    print("Detailed classification report:")
-    print()
-    print("The model is trained on the full development set.")
-    print("The scores are computed on the full evaluation set.")
-    print()
-    y_true, y_pred = y_test, clf.predict(X_test)
-    print(classification_report(y_true, y_pred))
-    print()
-
-# Note the problem is too easy: the hyperparameter plateau is too flat and the
-# output model is the same for precision and recall with ties in quality.
+# %%
+# We can also define a function that will implement the strategy to use to
+# select the best candidate from the `cv_results_` returns from the
+# :class:`~sklearn.model_selection.GridSearchCV`. Once the candidate selected,
+# it will be automatically refitted by the
+# :class:`~sklearn.model_selection.GridSearchCV` instance. Here, the strategy
+# is to short list the model which are the best in terms of precision and
+# recall. From the selected models, we finally select the fastest model at
+# predicting.
```

Comment:
I think using the F1-score directly allows selecting the best model in term of precision and recall as it is their harmonic mean. The strategy bellow proceeds in two steps (filtering on the precision and then on the recall), but would it be simpler to adapt it to solely use the F1-score? Would this be equivalent?